File: src/graph_builder.cpp
#include "graph_builder.hpp"

#include <iostream>


GraphBuilder::~GraphBuilder() {
    for (auto node : nodes) delete node;
    for (auto edge : edges) delete edge;
}

void GraphBuilder::addNode(Node* node) {
    nodes.push_back(node);
}

void GraphBuilder::addEdge(Edge* edge) {
    edges.push_back(edge);
}


void GraphBuilder::calculateAdjacencyMatrix(std::vector<std::vector<int>>& adjacencyMatrix, bool sym) {
    // Create a mapping of node IDs to indices
    std::unordered_map<int, int> nodeMapping = createNodeMapping();

    // Clear the matrix before populating it again
    adjacencyMatrix.clear();
    adjacencyMatrix.resize(nodes.size(), std::vector<int>(nodes.size(), 0));

    // Populate the adjacency matrix using edge lengths
    for (const auto& edge : edges) {
        // Check if the from and to nodes exist in the node mapping
        if (nodeMapping.find(edge->from->id) == nodeMapping.end() ||
            nodeMapping.find(edge->to->id) == nodeMapping.end()) {
            // Handle error: Nodes not found in the node mapping
            std::cerr << "Error: Nodes not found in the node mapping." << std::endl;
            return;
        }

        // Get the indices of the from and to nodes
        int fromId = nodeMapping[edge->from->id];
        int toId = nodeMapping[edge->to->id];

        // Assign the edge length to the corresponding matrix entry
        adjacencyMatrix[fromId][toId] = edge->length;

        // If symmetric, assign the same length to the corresponding entry
        if (sym) {
            adjacencyMatrix[toId][fromId] = edge->length;
        }
    }
}

void GraphBuilder::printAdjacencyMatrix(const std::vector<std::vector<int>>& adjacencyMatrix) {
    std::cout << "Adjacency Matrix:" << std::endl;
    for (const auto& row : adjacencyMatrix) {
        for (int val : row) {
            std::cout << val << " ";
        }
        std::cout << std::endl;
    }
}

std::unordered_map<int, int> GraphBuilder::createNodeMapping() {
    std::unordered_map<int, int> mapping;
    int index = 0;
    for (const auto& node : nodes) {
        mapping[node->id] = index;
        node->mappedId = index;
        ++index;
    }
    return mapping;
}

void GraphBuilder::printNodeMapping(const std::unordered_map<int, int>& nodeMapping) {
    std::cout << "Node Mapping:" << std::endl;
    for (const auto& pair : nodeMapping) {
        std::cout << "Original ID: " << pair.first << " -> Sequential Index: " << pair.second << std::endl;
    }
}

File: python_scripts/graph_builder.py
#!/usr/bin/env python3
import math

class Edge:
    def __init__(self, from_node_id, to_node_id, type):
        self.from_node_id = from_node_id
        self.to_node_id = to_node_id
        self.type = type
        self.length = 0  # Calculate later if needed

    def calculate_edge_length(self, graph):
        #print("from_node_id",self.from_node_id)
        for node in graph.nodes:
            if node.id==self.from_node_id:
                from_node=node
            elif node.id==self.to_node_id:
                to_node=node
        dx = to_node.x - from_node.x
        dy = to_node.y - from_node.y
        dz = to_node.z - from_node.z
        self.length = math.sqrt(dx ** 2 + dy ** 2 + dz ** 2)
        theta = math.atan2(dy, dx)  # Azimuthal angle
        phi = math.acos(dz / self.length) if self.length != 0 else 0  # Polar angle
        return(self.length, theta, phi)
    
class Graph:
    def __init__(self):
        self.nodes = []
        self.edges = []

    def __del__(self):
        for node in self.nodes:
            del node
        for edge in self.edges:
            del edge

    def add_node(self, node):
        self.nodes.append(node)

    def add_edge(self, edge):
        self.edges.append(edge)

    def calculate_adjacency_matrix(self):
        node_mapping = self.create_node_mapping()
        adjacency_matrix = [[0] * len(self.nodes) for _ in range(len(self.nodes))]

        for edge in self.edges:
            #print("from node")
            #print(edge.from_node_id.id)
            from_id = node_mapping[edge.from_node_id]
            to_id = node_mapping[edge.to_node_id]
            adjacency_matrix[from_id][to_id] = edge.length
            #adjacency_matrix[to_id][from_id] = edge.length  # Comment out for directed graphs

        return adjacency_matrix

    def print_adjacency_matrix(self, adjacency_matrix):
        print("Adjacency Matrix:")
        for row in adjacency_matrix:
            print(" ".join(map(str, row)))

    def create_node_mapping(self):
        node_mapping = {}
        for index, node in enumerate(self.nodes):
            node_mapping[node.id] = index
            node.mapped_id = index
        return node_mapping

    def print_node_mapping(self, node_mapping):
        print("Node Mapping:")
        for original_id, sequential_index in node_mapping.items():
            print(f"Original ID: {original_id} -> Sequential Index: {sequential_index}")

class Node:
    def __init__(self, id, x, y, z):
        self.id = id
        self.mapped_id = -1
        self.x = x
        self.y = y
        self.z = z
        self.parent = None
        self.children = []

    def add_child(self, child):
        self.children.append(child)
